main/lsp/requests/rename.cc

```diff
@@ -29,37 +30,135 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
                 fmt::format("Renaming constants defined in {} files is not supported; symbol {} is defined at {}",
                             filetype, symbol.data(gs)->name.show(gs), loc.filePosToString(gs));
             response->error = make_unique<ResponseError>((int)LSPErrorCodes::InvalidRequest, error);
-
             return false;
         }
     }
     return true;
 }
+
+// Replaces originalName with newName in the last dotted component if the string is of the form x.y.z; otherwise does a
+// simple ReplaceAll.
+std::string replaceLastDotted(std::string_view input, std::string_view originalName, std::string_view newName) {
+    if (absl::StrContains(input, ".")) {
+        std::vector<std::string> dotted = absl::StrSplit(input, ".");
+        // TODO: check for exactly one match
+        dotted[dotted.size() - 1] = absl::StrReplaceAll(dotted[dotted.size() - 1], {{originalName, newName}});
+        return absl::StrJoin(dotted, ".");
+    }
+    return absl::StrReplaceAll(input, {{originalName, newName}});
+}
+
+// Checks if s is a subclass of root, and updates isSubclass, visited, and subclasses vectors.
+bool updateSubclass(const core::GlobalState &gs, core::SymbolRef root, core::SymbolRef s, vector<bool> &isSubclass,
+                    vector<bool> &visited, vector<core::SymbolRef> &subclasses) {
+    if (visited[s.rawId()] == true) {
+        return isSubclass[s.rawId()];
+    }
+    visited[s.rawId()] = true;
+    if (s.rawId() == root.rawId()) {
+        subclasses.push_back(s);
+        isSubclass[s.rawId()] = true;
+        return true;
+    }
+    auto super = s.data(gs)->superClass();
+    if (super.exists()) {
+        if (updateSubclass(gs, root, super, isSubclass, visited, subclasses)) {
+            subclasses.push_back(s);
```

Comment:
I don't see how it duplicates results? The function can't ever do line 59 and this line for the same class because it tracks `visited`.  If we move out the vector from this function we'll add a factor of `class hierarchy length` to the overall time complexity, I think? Because we'll have to walk up the path from class to root for every class we visit, e.g. if we have `A < B < C`, we'll have to walk the B->C link twice if we happen to visit A before B in the outer loop.